lighthouse-core/report/html/renderer/details-renderer.js

```diff
@@ -315,41 +315,94 @@ class DetailsRenderer {
    * @return {LH.Audit.Details.OpportunityColumnHeading}
    */
   _getCanonicalizedHeading(heading) {
-    let subRows;
-    if (heading.subRows) {
+    let subHeading;
+    if (heading.subHeading) {
       // @ts-ignore: It's ok that there is no text.
-      subRows = this._getCanonicalizedHeading(heading.subRows);
-      if (!subRows.key) {
+      subHeading = this._getCanonicalizedHeading(heading.subHeading);
+      if (!subHeading.key) {
         // eslint-disable-next-line no-console
         console.warn('key should not be null');
       }
-      subRows = {...subRows, key: subRows.key || ''};
+      subHeading = {...subHeading, key: subHeading.key || ''};
     }
 
     return {
       key: heading.key,
       valueType: heading.itemType,
-      subRows,
+      subHeading,
       label: heading.text,
       displayUnit: heading.displayUnit,
       granularity: heading.granularity,
     };
   }
 
   /**
-   * @param {LH.Audit.Details.ItemValue[]} values
-   * @param {LH.Audit.Details.OpportunityColumnHeading} heading
-   * @return {Element}
+   * @param {LH.Audit.Details.OpportunityItem | LH.Audit.Details.TableItem} item
+   * @param {(LH.Audit.Details.OpportunityColumnHeading | null)[]} headings
    */
-  _renderSubRows(values, heading) {
-    const subRowsElement = this._dom.createElement('div', 'lh-sub-rows');
-    for (const childValue of values) {
-      const subRowElement = this._renderTableValue(childValue, heading);
-      if (!subRowElement) continue;
-      subRowElement.classList.add('lh-sub-row');
-      subRowsElement.appendChild(subRowElement);
+  _renderTableRow(item, headings) {
+    const rowElem = this._dom.createElement('tr');
+
+    for (const heading of headings) {
+      const value = heading && heading.key && item[heading.key];
+
+      let valueElement;
+      if (heading && heading.key !== null && value !== undefined && value !== null) {
+        valueElement = this._renderTableValue(value, heading);
+      }
+
+      if (heading && valueElement) {
+        const classes = `lh-table-column--${heading.valueType}`;
+        this._dom.createChildOf(rowElem, 'td', classes).appendChild(valueElement);
+      } else {
+        // Empty cell is rendered for a column if:
+        // - the pair is null
+        // - the heading key is null
+        // - the value is undefined/null
+        this._dom.createChildOf(rowElem, 'td', 'lh-table-column--empty');
+      }
```

Comment:
to be able to skip the `heading && ...` checks, what about breaking the empty cell check in two so the no heading case can be ruled out early? Something like
```js
for (const heading of headings) {
  // Empty cell if no heading or heading key for this column.
  if (!heading || !heading.key) {
    this._dom.createChildOf(rowElem, 'td', 'lh-table-column--empty');
    continue;
  }

  const value = item[heading.key];
  const valueElement = this._renderTableValue(value, heading);

  if (valueElement) {
    const classes = `lh-table-column--${heading.valueType}`;
    this._dom.createChildOf(rowElem, 'td', classes).appendChild(valueElement);
  } else {
    // Empty cell if there's no value to put in it.
    this._dom.createChildOf(rowElem, 'td', 'lh-table-column--empty');
  }
}
```